### PR TITLE
Added a proper support for `cargo --message-format json`

### LINKED ISSUE #77: Switching between terminal and Atom with linter-rust causes a clean rebuild
This causes long rebuilds of a full project.
However, if I remain in one environment (say `linter-rust`) clean rebuilds do not happen.

This was introduced recently, probably, with my change of how environment is passed to `cargo`. 

I will debug this and try to fix, but I'm curious if I'm the only one who encountered this.
Please upvote this or leave a response if you have experienced such an issue.

### PR DESCRIPTION
**Pull request by White-Oak**
Massively refactored code, so the error mode behaviour is less dependent on conditions and more polymorphic.

I've tested this in all five modes that `linter-rust` currently supports:
* Old rustc plain-text messages
* Old cargo plain-text messages
* Rustc json errors
* Intermediate cargo json errors, enabled by `RUSTFLAGS`, that caused several problems and only enabled for a short ammount of time from AUgust to October 2016
* Proper cargo json errors enabled by `--message-format json`, that are introduced by this PR

In short, I split the linter in two files: `linter-rust.coffee` and `mode.coffee`. The latter is now responsible on mode-specific interactions such as error parsing, arguments generating. 

Fixes #77.

As far as I tested it, it works, but I, probably, need some review on code.

### CODE REVIEW

**Comment by White-Oak** [comment]
@Arcanemagus if you have time and wish, can you please review code I wrote, just as you did at #74?

**Comment by Arcanemagus** [comment]
Sorry, just saw this. Looking over it now!

File: lib/linter-rust.coffee [hunk 1]
```diff
@@ -273,19 +140,32 @@ class LinterRust
           result.push ['--cfg', "feature=\"#{f}\""]
         result
 
-  ableToJSONErrors: (curDir) =>
+  decideErrorMode: (curDir, commandMode) =>
     # current dir is set to handle overrides
-
     sb_exec.exec(@rustcPath, ['--version'], {stream: 'stdout', cwd: curDir, stdio: 'pipe'}).then (stdout) =>
```
Comment (Arcanemagus) [inline lib/linter-rust.coffee]: The same can be said for other "check calls", such as [here](https://github.com/AtomLinter/linter-rust/pull/84/files#diff-b6440b005dd344507743aaaa07d74383R168)

File: lib/mode.coffee [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+buildCargoArguments = (linter, cargoManifestPath) ->
+  cargoArgs = switch linter.cargoCommand
+    when 'check' then ['check']
+    when 'test' then ['test', '--no-run']
+    when 'rustc' then ['rustc', '-Zno-trans', '--color', 'never']
+    when 'clippy' then ['clippy']
+    else ['build']
+
+  compilationFeatures = linter.compilationFeatures(true)
+  buildCargoPath(linter.cargoPath).then (cmd) ->
```
Comment (Arcanemagus) [inline lib/mode.coffee]: 🐛 I think you missed a `linter.cargoCommand` as a second argument to this function. As it is currently written since you aren't providing a `cargoCommand` here, this will _always_ return `[cargoPath]`, so might as well just remove the entire `buildCargoPath` function as this is the only location it is being called from.

If the `buildCargoPath` function is being removed, so should the `usingMultitoolForClippy` as it is only used in there.

File: lib/mode.coffee [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    else
+      [cargoPath]
+
+usingMultitoolForClippy = () ->
+  # Try to use rustup
+  sb_exec.exec 'rustup', ['--version'], {ignoreExitCode: true}
+    .then ->
+      result: true, tool: 'rustup'
+    .catch ->
+      # Try to use odler multirust at least
```
Comment (Arcanemagus) [inline lib/mode.coffee]: 🐛 "odler" -> "older"

File: lib/mode.coffee [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+pattern = XRegExp('(?<file>[^\n\r]+):(?<from_line>\\d+):(?<from_col>\\d+):\\s*\
+  (?<to_line>\\d+):(?<to_col>\\d+)\\s+\
+  ((?<error>error|fatal error)|(?<warning>warning)|(?<info>note|help)):\\s+\
+  (?<message>.+?)[\n\r]+($|(?=[^\n\r]+:\\d+))', 's')
+
+parseOldMessages = (output, disabledWarnings) ->
+  elements = []
+  XRegExp.forEach output, pattern, (match) ->
+    if match.from_col == match.to_col
+      match.to_col = parseInt(match.to_col) + 1
```
Comment (Arcanemagus) [inline lib/mode.coffee]: 🐛 `parseInt` -> `Number.parseInt(foo, 10)`

Comment (Arcanemagus) [inline lib/mode.coffee]: 💡 Situations like this are where [`atom-linter::rangeFromLineNumber`](https://github.com/steelbrain/atom-linter/blob/v8.0.0/src/index.js#L16) is great as it will automatically determine a range highlighitng the next "word", _as defined by the language_. So you can feed it a single point and it will generate a better range than just the single character. It would require some refactoring though as it requires access to the `TextEditor` of the file being linted.

File: lib/mode.coffee [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+      message: match.message
+      file: match.file
+      range: range
+    elements.push element
+  buildMessages elements, disabledWarnings
+
+parseJsonMessages = (messages, disabledWarnings) ->
+  elements = []
+  for input in messages
+    continue unless input and input.spans
```
Comment (Arcanemagus) [inline lib/mode.coffee]: 📝 [Linter messages](https://github.com/steelbrain/linter/wiki/Linter-API#messages) don't (currently) require a range, is this ignoring potentially useful messages just because they don't have a span?

File: lib/mode.coffee [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    primary_span = input.spans.find (span) -> span.is_primary
+    continue unless primary_span
+    range = [
+      [primary_span.line_start - 1, primary_span.column_start - 1],
+      [primary_span.line_end - 1, primary_span.column_end - 1]
+    ]
+    input.level = 'error' if input == 'fatal error'
```
Comment (Arcanemagus) [inline lib/mode.coffee]: 🐛 Should `input == 'fatal error'` be `input.message == 'fatal error'`?

File: lib/mode.coffee [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+      buildRustcArguments(linter, file).then (cmd_res) ->
+        [file, cmd] = cmd_res
+        cmd = cmd.concat ['--error-format=json']
+        [file, cmd]
+
+  JSON_CARGO:
+    neededOutput: (stdout, stderr) ->
+      stdout
+
+    parse: (output, disabledWarnings) ->
```
Comment (Arcanemagus) [inline lib/mode.coffee]: 🐛 Can the functionality here be merged into `parseJsonOutput`? They are _almost_ identical.

File: lib/mode.coffee [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+      message.trace.push
+        type: "Trace"
+        text: children.message
+        filePath: element.file
+        range: children.range or element.range
+  message
+
+buildRustcArguments = (linter, paths) ->
+  [editingFile, cargoManifestPath] = paths
+  Promise.resolve().then () =>
```
Comment (Arcanemagus) [inline lib/mode.coffee]: 🐛 It looks like you are trying to use this as a way to return a Promise, you should just go directly there:
`new Promise (resolve, reject) =>`

I'd consider this an optional change though I guess, as it makes the CoffeeScript of the function even worse than it is from the start 😛.

File: lib/mode.coffee (hunk 4 above)
Comment (White-Oak) [inline lib/mode.coffee]: @Arcanemagus if I add `atom-linter` as a dependency, I should probably remove `sb-exec`, since it's included, should I?

**Comment by White-Oak** [comment]
@Arcanemagus huge thanks for the big and helpful review!

The tricky part, although, is caching results, as a user might switch tools `on the fly`, I've added a simple caching wrapper with TTL, that survives only `TTL` lints.

File: lib/mode.coffee (hunk 4 above)
Comment (Arcanemagus) [inline lib/mode.coffee]: Correct, `atom-linter::exec` is a wrapper for `sb-exec` so there is no need to include it directly.

**Comment by Arcanemagus** [comment]
That's why that was just a suggestion and not something that needed to be addressed. As I said it would work just fine as is 😉.

**Comment by Arcanemagus** [comment]
Hmm, so the way I was caching it in other providers I cached the version based on the path to the executable, I was thinking of something along the lines of caching the result here based on the current directory as that should handle changing executables by directory which is what most "executable version management" systems do.

Would that not work here? Or does the version get changed often for the same directory?

**Comment by zofrex** [comment]
I hope it's not rude to butt in, I was trying to find a solution for my cargo builds being invalidated all the time, stumbled across this PR, and did some unsolicited testing :)

File: lib/mode.coffee [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+        [file, cmd] = cmd_res
+        cmd = cmd.concat ['--error-format=json']
+        [file, cmd]
+
+  JSON_CARGO:
+    neededOutput: (stdout, stderr) ->
+      stdout
+
+    parse: (output, options) ->
+      options.additionalFilter = (json) ->
```
Comment (zofrex) [inline lib/mode.coffee]: I think this is a bug, on my machine it was filtering out 100% of errors returned from the compiler. I think that either "json" here should be "input" to match below, or vice versa.

File: lib/linter-rust.coffee [hunk 10]
```diff
@@ -273,19 +141,43 @@ class LinterRust
...
+      atom_linter.exec(@rustcPath, ['--version'], {cwd: curDir}).then (stdout) =>
+        try
+          match = XRegExp.exec(stdout, @patternRustcVersion)
+          if match
+            nightlyWithJSON = match.nightly and match.date > '2016-08-08'
+            stableWithJSON = not match.nightly and semver.gte(match.version, '1.12.0')
+            canUseIntermediateJSON = nightlyWithJSON or stableWithJSON
+            switch commandMode
+              when 'cargo'
+                canUseProperCargoJSON = match.nightly and match.date > '2016-10-10'
```
Comment (White-Oak) [inline lib/linter-rust.coffee]: You are right, the right starting version is probably 10-08 or 10-09 since it was merged 15 days ago in  cargo. Thanks for the help, I 'll get this fixed :)

**Comment by White-Oak** [comment]
@Arcanemagus well, it's not that _common_ to change toolchains in Rust, but it's something that one can and expect being able to easily do. For example, consider someone coding a library with a stable compiler in Atom with linter-rust, and then switching over to a beta or a nightly toolchain (that's where the override is being set for a project, and toolchains' versions are changed as a consequence) to check if there are any warnings considering any possible stabilizations or deprecations. 

> Hmm, so the way I was caching it in other providers I cached the version based on the path to the executable, I was thinking of something along the lines of caching the result here based on the current directory as that should handle changing executables by directory which is what most "executable version management" systems do.
> 
> Would that not work here? Or does the version get changed often for the same directory?

There is, actually, no precedent of 'per-directory' toolchain swapping in Rust, as far as I'm aware of. 
In Rust you can, of course, set an overriding toolchain for every directory you'd like, but that's a _really rare_ use case. Most of work in Rust is done on a 'per-project' basis, so setting an override for some subdirectory of that project does not really make much sense in language ecosystem terms. 

On the other side, I can imagine someone making a large directory of subdirectories that have one `main.rs` each and not using Rust's package manager (`cargo`), but `rustc` compiler directly to compile and run them. This is where a more fine-grained caching of toolchains' versions can be useful, as one can set overrides on several subdirectories, while not setting that override on others.
However, the ability to cache versions 'per-directory' can be added to any caching mechanism, as long as it exists, that's something that is need to be reconsidered.

All in all, there are several ways to handle caching:

1. Don't cache at all. That's simple and that's how it works now in `master`. The downside is increased performance hit per lint action.
2. Cache 'forever' (as in 'until editor is closed'). This is simple, and I think, what should be actually done, but thee may be an issue of people switching toolchains -- that would force them to reopen Atom every time
   1. Maybe there is a point in introducing an option in `rust-linter`, say, `Cache toolchains' versions`, disabling which would force code to check versions every time.
3. Cache using several heuristics like the caching system in this PR that uses TTL. When I wrote that system, initially I thought it was really great and would solve defined above issues, but now I understand, that it's not a real solution, instead it's being harder to reason about for a user (e.g. "why does `linter-rust` not lint after I changed a toolchain's version, but then, suddenly starts to lint?").

What do you think?

**Comment by Arcanemagus** [comment]
I would vote for option 2, since it sounds unlikely that it would change during a session of Atom being open, at least that's the reasoning I went with for `linter-phpcs` 😛.

Maybe a note in the readme that if the use changes the version a restart of Atom will be required to catch any additional features (un)available in the new version?

Just wanted to say thanks for all the work you've been putting into this provider btw!

**Comment by colin-kiegel** [comment]
I've been following this discussion. :-)

I think switching the toolchain during a session is not so rare. I, for example, recently worked on a rust macro library which is supposed to run on rust-stable. However all the nice macro-debugging tools are only available in rust-nightly. Whenever I discovered a hard bug with rust-stable I switched to rust-nightly until I fixed it - and back again. All during one atom session.

I think option 2.1 would be a good start, i.e. let the user decide if caching is enabled. From this point on any kind of heuristic can still be added in the future.

**Comment by zofrex** [comment]
What is the performance hit of not caching?

**Comment by White-Oak** [comment]
@zofrex Well, to be honest, it's not that big comparing to compilation times :) But It's more a future-proofing thing. Rustc is constantly improving and we are about to have incremental compilation soon. As an example of compilation times improving see [this issue](https://github.com/rust-lang/rust/issues/31157#issuecomment-255389265) (in short, the compilation time went from 30 secs to 5 in a time span of year).

Going into details, `linter-rust` calls two/three shell commands per lint process (`rustup --version`, `multirust --version` if you have no rustup installed, and `rustc --version`). Calling shell commands means starting a new process etc and is often compared to performing an IO operations -- it's better to avoid them.

**Comment by White-Oak** [comment]
@Arcanemagus it's me who should thank you for your reviews and advices! (: Thanks for helping me out here and there.

Kinda off-topic, but I'd ask you another linter related question. I understand that most linters do not have big timings on performing a lint, but as you've probably seen in the discussions Rust does suffer from big compilation times.  
And considering that, have there ever been a proposal on swapping this (that's at the bottom of the editor):
spectacle lh2471
with a 'Linting' label while the linting is in process? If yes, why has it been turned down? If no, do you think it makes sense to propose that one? Since it would really help `linter-rust`'s users, since it is sometimes really hard to understand, if a linting is undergoing, or is there _something wrong_.

**Comment by White-Oak** [comment]
@colin-kiegel thanks for the input! Yes, I think it's what we should go with. I've been reluctant to add another option, since I know people are sometimes confused with a number of options, but I hope 10 options is not a really big deal..

**Comment by White-Oak** [comment]
I've enabled 'caching forever' to be a default choice with an option in settings called `Allowed To Cache Versions`. 

@Arcanemagus if you have no other notices, I believe you can mark your review as a positive. Thanks again for your review.

**Comment by Arcanemagus** [comment]
@White-Oak Linter v2 incorporates the busy-signal package, which has a nice indicator when it is currently running a lint: image

In fact, it completely removes the "No issues" text. you can check it out following the instructions here: https://github.com/steelbrain/linter-ui-default

**Comment by White-Oak** [comment]
@gmist I hope you are not against that I merged this.

**Comment by gmist** [comment]
of course not against

**Comment by colin-kiegel** [comment]
PS: I would suggest to default to 'no caching' to be on the safe side.
People who want more juice will probably take some time tweaking settings
anyway.

Oak <notifications@github.com> schrieb am Mo., 24. Okt. 2016, 22:36:

**Comment by White-Oak** [comment]
@colin-kiegel default settings are rarely changed, if making 'no caching' be the default choice, then why use caching anyway.

**Comment by colin-kiegel** [comment]
@White-Oak I guess in the end it's a matter of taste. Its ok for me to change the setting, but I would still suggest at least add a little warning in the settings description, that the rust linter will not detect a change in toolchain until restart. Again of course it's your choice. 

But this discussion reminds me, there may be another "hacky" approach instead of caching. You mentioned that you want to avoid spawning processes for shell commands. It may be possible to exchange a handful of small shell calls plus logic and bundle them into a small shell script instead, which can be passed to only one shell call. Of course the small shell script needs to be interpreted by the shell, but the spawning of processes can be saved. However the details of this approach will most likely be very OS-specific. If it is possible at all it would probably be a bit brittle - but I still wanted to share the thought, just in case... ^^

PS: Thanks for the great work.

**Comment by White-Oak** [comment]
@colin-kiegel there is one :)
spectacle lh1872

**Comment by AaronM04** [comment]
Maybe this isn't the right place to ask, but:

Are the changes in this PR present in version 0.7.0? I am still experiencing the #77 problem with rust nightly from yesterday, and version 0.7.0, with Atom 1.11.2 on Linux x86_64. I just discovered a workaround, which is to compile with `--release` in the terminal.

**Comment by White-Oak** [comment]
@AaronM04 thanks for the report! This is really weird, and really strange! Can you please as well post a version of rustc?
I will try my best to investigate that!

**Comment by AaronM04** [comment]
@White-Oak 
```
$ rustc --version
rustc 1.12.1 (d4f39402a 2016-10-19)
```

**Comment by AaronM04** [comment]
@White-Oak 
Looks like I wasn't using nightly after all.

**Comment by White-Oak** [comment]
@AaronM04 yes, it is a stable from when JSON errors were only supported via RUSTCFLAGS. Hope it cleared everything out. Latest stable (1.13), beta and nightly version all support proper JSON errors' support. Hope it helps :)

**Comment by AaronM04** [comment]
@White-Oak I switched to the real nightly and everything works. Thanks!
